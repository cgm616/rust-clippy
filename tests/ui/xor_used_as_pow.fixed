// run-rustfix
#![warn(clippy::xor_used_as_pow)]
#![allow(clippy::identity_op)]

// Should not be linted
#[allow(dead_code)]
enum E {
    First = 1 ^ 8,
    Second = 2 ^ 8,
    Third = 3 ^ 8,
    Tenth = 10 ^ 8,
}

fn main() {
    // These should succeed:
    let _ = 9 ^ 3; // lhs other than 2 or 10
    let _ = 0x02 ^ 6; // lhs not decimal
    let _ = 2 ^ 0x10; // rhs hexadecimal
    let _ = 10 ^ 0b0101; // rhs binary
    let _ = 2 ^ 0o1; // rhs octal
    let _ = 10 ^ -18; // negative rhs

    // These should fail
    let _ = 1_u32 << 3;
    let _ = 10 ^ 4;
    let _ = 1_u64 << 32;
    let _ = 1;
    let _ = 10 ^ 0;
    {
        let x = 15;
        let _ = 1 << x;
        let _ = 10 ^ x;
    }
}
