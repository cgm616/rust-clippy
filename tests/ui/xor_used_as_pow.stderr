error: it appears you are trying to get a power of two, but `^` is not an exponentiation operator
  --> $DIR/xor_used_as_pow.rs:24:13
   |
LL |     let _ = 2 ^ 3;
   |             ^^^^^ help: use a bitshift or constant instead: `1_u32 << 3`
   |
   = note: `-D clippy::xor-used-as-pow` implied by `-D warnings`

error: `^` is not an exponentiation operator but appears to have been used as one
  --> $DIR/xor_used_as_pow.rs:25:13
   |
LL |     let _ = 10 ^ 4;
   |             ^^^^^^
   |
   = help: did you mean to use .pow()?

error: it appears you are trying to get a power of two, but `^` is not an exponentiation operator
  --> $DIR/xor_used_as_pow.rs:26:13
   |
LL |     let _ = 2 ^ 32;
   |             ^^^^^^ help: use a bitshift or constant instead: `1_u64 << 32`

error: it appears you are trying to get a power of two, but `^` is not an exponentiation operator
  --> $DIR/xor_used_as_pow.rs:27:13
   |
LL |     let _ = 2 ^ 0;
   |             ^^^^^ help: use a bitshift or constant instead: `1`

error: `^` is not an exponentiation operator but appears to have been used as one
  --> $DIR/xor_used_as_pow.rs:28:13
   |
LL |     let _ = 10 ^ 0;
   |             ^^^^^^
   |
   = help: did you mean to use .pow()?

error: it appears you are trying to get a power of two, but `^` is not an exponentiation operator
  --> $DIR/xor_used_as_pow.rs:31:17
   |
LL |         let _ = 2 ^ x;
   |                 ^^^^^ help: use a bitshift or constant instead: `1 << x`

error: `^` is not an exponentiation operator but appears to have been used as one
  --> $DIR/xor_used_as_pow.rs:32:17
   |
LL |         let _ = 10 ^ x;
   |                 ^^^^^^
   |
   = help: did you mean to use .pow()?

error: aborting due to 7 previous errors

